src/sentry/releases/servicer.py

```diff
@@ -0,0 +1,161 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from dataclasses import dataclass
+from datetime import datetime, timedelta, timezone
+from typing import TYPE_CHECKING, Any, DefaultDict, Dict, List, Optional, TypedDict
+
+from sentry.api.serializers import serialize
+from sentry.releases.repository import ReleaseThresholdsRepository, TimeRange
+from sentry.services.hybrid_cloud.organization import RpcOrganization
+
+if TYPE_CHECKING:
+    from sentry.models.deploy import Deploy
+    from sentry.models.organization import Organization
+    from sentry.models.project import Project
+    from sentry.models.release import Release
+    from sentry.models.release_threshold.release_threshold import ReleaseThreshold
+    from sentry.models.releaseprojectenvironment import ReleaseProjectEnvironment
+
+
+class SerializedThreshold(TypedDict):
+    date: datetime
+    environment: Dict[str, Any] | None
+    project: Dict[str, Any]
+    release: str
+    threshold_type: int
+    trigger_type: str
+    value: int
+    window_in_seconds: int
+
+
+class EnrichedThreshold(SerializedThreshold):
+    end: datetime
+    is_healthy: bool
+    key: str
+    project_slug: str
+    project_id: int
+    start: datetime
+    metric_value: int | None
+
+
+@dataclass
+class FlattenedThresholds:
+    # TODO: this can probably be improved to a better data structure as the keys for both fields match
+    thresholds_by_type: DefaultDict[int, Dict[str, List]]
+    query_windows_by_type: DefaultDict[int, Dict[str, datetime]]
+
+
+class ReleaseThresholdServicer:
+    def __init__(self, repository: Optional[ReleaseThresholdsRepository] = None) -> None:
+        self.repository = repository if repository else ReleaseThresholdsRepository()
+
+    @classmethod
+    def _get_latest_deploy_for_release_threshold(
+        cls, threshold: ReleaseThreshold, release: Release, project: Project
+    ) -> Optional[Deploy]:
```

Comment:
Looking at it now, this should probably be moved to the repository layer, as it's mostly just doing DB related checks and logic